lib/backend/test/atomicwrite_shim.go

```diff
@@ -0,0 +1,207 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package test
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+// RunBackendComplianceSuiteWithAtomicWriteShim runs the old backend compliance suite against the provided backend
+// with a shim that converts all calls to single-write methods (all write methods but DeleteRange) into calls to
+// AtomicWrite. This is done to ensure that the relationship between the conditional actions of AtomicWrite and the
+// single-write methods is well defined, and to improve overall coverage of AtomicWrite implementations via reuse.
```

Comment:
An optimized backend implementation is going to fall back to the dedicated single-operation implementations if a single action is passed. The shim should probably always add some check against the nonexistence of some nonce key, as an "effortful" noop.